### PR TITLE
Replaying cost match tests

### PR DESCRIPTION
**Pull request by zsluedem**
## Overview

Previously we disable replay cost tests because we  got some problems on consume on the same channel https://github.com/rchain/rchain/pull/3579/files#diff-a8950e05083ebb606328a9eb162794f4d131f122e08a886f90c911e77969fd82L252 . But we fixed that in https://github.com/rchain/rchain/pull/2888.
So it would be better to re-enable these tests.

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala [hunk 1]
```diff
@@ -263,6 +264,22 @@ class CostAccountingSpec extends FlatSpec with Matchers with PropertyChecks with
     }
   }
 
+  "replay the same term with not enough phlo" should "get the same result" in {
+    val r = scala.util.Random
+    for (_ <- 1 to 100) {
+      val long        = ((r.nextLong % 0X144000000L) + 0X144000000L) % 0X144000000L
```
Comment (nzpr) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala]: I'm curious what this is doing?

**Review by nzpr — APPROVED** [review]
LGTM, minor question

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala (hunk 1 above)
Comment (zsluedem) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala]: To be very honest, I don't know. :) Maybe it is trying to generate a number with some kind of range.

Comment (zsluedem) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala]: Hm. I think it is useless. I would just remove it.

Comment (zsluedem) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala]: After https://github.com/rchain/rchain/runs/4932655784?check_suite_focus=true, I think these codes are for the `r` is negative situation.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala [hunk 2]
```diff
@@ -249,24 +250,10 @@ class CostAccountingSpec extends FlatSpec with Matchers with PropertyChecks with
       }
     }
 
-  // TODO: Remove ignore when bug RCHAIN-3917 is fixed.
-  it should "be repeatable when generated" ignore {
+  "replay the same term with same event logs" should "get the same cost" in {
```
Comment (tgrospic) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala:253]: We don't want to remove existing test, it's testing bigger scope of Rholang terms.

Please add this as an additional test which is testing much smaller scope.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala [hunk 3]
```diff
@@ -171,7 +172,7 @@ class CostAccountingSpec extends FlatSpec with Matchers with PropertyChecks with
             remainder /= 4
             joins += f"_ ${arrow} @${ch}"
           }
-          val joinStr = joins.mkString("; ")
+          val joinStr = joins.toSet.mkString("; ")
```
Comment (tgrospic) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala:175]: This function is generator of Rholang terms. Each index number corresponds to different terms so conversion to Set means to make some of the indexes equal.

What is the reason for this change?

Comment (zsluedem) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala:175]: https://rchain.atlassian.net/browse/RCHAIN-3917?
Rholang doesn't support it.
Same joins means `consuming two messages from the same channel`

Comment (tgrospic) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala:175]: > Every number gets decoded into a unique term

If we change internal behavior of this function, every number will not produce a unique term.
So to get terms without joins on the same channel these numbers must be skipped.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala (hunk 2 above)
Comment (zsluedem) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala:253]: No tests are removed . Why did you say that? I just rephrased it.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala (hunk 3 above)
Comment (zsluedem) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala:175]: Could you elaborate?
What specific do you want?

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala [hunk 4]
```diff
@@ -277,6 +264,22 @@ class CostAccountingSpec extends FlatSpec with Matchers with PropertyChecks with
     }
   }
 
+  "replay the same term with not enough phlo" should "get the same result" in {
+    val r = scala.util.Random
+    for (_ <- 1 to 100) {
```
Comment (tgrospic) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala:269]: Here are many tests removed because only 100 terms are considered. Each number represents a unique term so reducing the number of iterations means skipping tests for all these Rholang terms.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala (hunk 3 above)
Comment (tgrospic) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala:175]: If you want to skip some Rholang terms it's better to find the range of numbers which generates these terms and not used them in tests.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala (hunk 2 above)
Comment (tgrospic) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala:253]: I've added comment in the next test to explain.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala (hunk 3 above)
Comment (zsluedem) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala:175]: Is it realistic to do that? Is it necessary? 
What is the shortcoming of the current solution? 
And how to get those numbers? Iterate all numbers and hardcode exclusions here?

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala (hunk 4 above)
Comment (zsluedem) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala:269]: If you look at the generated terms, they are almost identical with little changes. :)

Comment (zsluedem) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala:269]: Not neccessary to do all for 10000 times to make the test run much longer.

Comment (tgrospic) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala:269]: Yes, that's the point of this test, to check all possible combinations with very small deviations. So I'd say that running first 100 is not necessary at all because these basic terms are already covered by Rholang existing tests.

Maybe we can add another generator with terms without same channels in joins. But in a way to not just collapse joins on the same channels because that means we will generate and run the same terms for different input numbers.
